File: algorithm/stack/MinStack/155-MinStack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, value):
        self.stack.append(value)

        if len(self.minStack) == 0 or value <= self.getMin():
            self.minStack.append(value)

    def pop(self):
        value = self.stack.pop()
        if value == self.getMin():
            self.minStack.pop()

    def top(self):
        return self.stack[len(self.stack - 1)]
    
    def getMin(self):
        return self.minStack[len(self.minStack) - 1]

File: algorithm/stack/MinStack/155-MinStack.ts
class MinStack<T>{

    constructor(
        private stack: T[] = [],
        private minStack: T[] = []
    ){}

    push(value:T){
        this.stack.push(value)
        if(this.stack.length === 0 || value <= this.getMin()){
            this.minStack.push(value)
        }
    }

    pop(){
        const value = this.stack.pop()
        if(value === this.getMin){
            this.minStack.pop()
        }
    }

    top(){
        return this.stack[this.stack.length - 1]
    }

    getMin(){
        return this.minStack[this.minStack.length - 1]
    }
}
//  O(1) SC
//  O(n) TC